keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -105,21 +91,48 @@ flexibility will allow us to experiment with approaches for data populators.
 - Enable users to create pre-populated volumes in a manner consistent with current practice
 - Enable developers to innovate with new an interesting data sources
 - Avoid changing existing workflows and breaking existing tools with new behavior
+- Ensure that users continue to get accurate error messages and feedback about volume
+  creation progress, whether using existing methods of volume creation or specifying a
+  data source associated with a volume populator.
 
 ### Non-Goals
 
 - This proposal DOES NOT recommend any specific approach for data populators. The specific
   design of how a data populator should work will be handled in a separate KEP.
-- We do not propose any new object types. Specifically, no special "populator"
-  object type. Populators should function based on their own object types.
 
 ## Proposal
 
-The validation for `DataSource` field should be removed entirely, in the fullness of time.
-
-Short term, we should add a new alpha feature gate, which allows arbitrary objects to be
-specified for the `DataSource` field, with the intention to eventually make this behavior
-the standard. 
+Validation for the `DataSource` field should be moved from the core to a validation
+webhook. Because the existing validation only allow snapshots and volumes as data sources,
+a feature gate is added to relax that validation, allowing the existing object types plus
+any `Kind` of CR. Each volume populator should be able to register one of more kinds of
+objects that it understands.
+
+In order to ensure that dead-on-arrival PVCs (ones that will never be bound) can't be
+created, the webhook needs to deny creation of PVCs with a `DataSource` `Kind` that it
+doesn't recognize. This ensures that PVCs which do get created will have some controller
+responsible for handling dynamic provisioning, which can either create the volume as
+requested, or provide useful error message events on the PVC to give the user feedback.
+
+To support the registration of valid kinds, we add a new CRD called VolumePopulator. For
+example:
+
+```
+kind: VolumePopulator
+apiVersion: populator.storage.k8s.io/v1alpha1
+metadata:
+  name: example-populator
+sourceKind:
+  group: example.storage.k8s.io
+  kind: Example
+```
+
```

Comment:
Can you describe where this new CRD and its associated controller will live?  Will this be under external-provisioner repo or a separate repo?

Reply:
A side question, can we register more than one populator for the same kind ?

Reply:
@xing-yang Here is my prototype: https://github.com/kubernetes-csi/external-provisioner/pull/491
I imagine it will live in external-provisioner, but that's up for debate.

@humblec That's a really good question I had not thought of. I think the short answer is no, but I realize that the way this CRD is written you could do that and it would work. I also recognize the necessity to have data sources that could be populated by any one of several implementation, and I want to tackle that use case within this framework, using just one CR.